pkg/distributor/distributor.go

```diff
@@ -310,11 +317,24 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 		}
 		stream.Entries = stream.Entries[:n]
 
+		var shardErr error
 		if d.cfg.ShardStreams.Enabled {
-			derivedKeys, derivedStreams := d.shardStream(stream, userID)
-			keys = append(keys, derivedKeys...)
-			streams = append(streams, derivedStreams...)
-		} else {
+			derivedKeys, derivedStreams, err := d.shardStream(stream, userID)
+			if err != nil {
+				level.Error(util_log.Logger).Log("msg", "error sharding stream", "err", err)
+				shardErr = err
+			} else {
+				keys = append(keys, derivedKeys...)
+				streams = append(streams, derivedStreams...)
+			}
+
+			if unshardableErr, ok := err.(*unshardableStreamErr); ok {
```

Comment:
I was considering this over the weekend. Returning an error is basically the same thing as a per-stream-rate limit error. What do you think about returning just increment a metric that can cause an alert?